internal/wal_verify_handler_test.go

```diff
@@ -0,0 +1,540 @@
+	currentSegment, _ := internal.NewWalSegmentDescription(currentSegmentName)
+
+	expectedIntegrityCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusOk,
+		Details: internal.IntegrityCheckDetails{
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000001",
+				EndSegment:    "000000050000000000000004",
+				SegmentsCount: 4,
+				Status:        internal.Found,
+			},
+			{
+				TimelineId:    6,
+				StartSegment:  "000000060000000000000005",
+				EndSegment:    "000000060000000000000008",
+				SegmentsCount: 4,
+				Status:        internal.Found,
+			},
+		},
+	}
+
+	expectedTimelineCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusOk,
+		Details: internal.TimelineCheckDetails{
+			CurrentTimelineId:        currentSegment.Timeline,
+			HighestStorageTimelineId: currentSegment.Timeline,
+		},
+	}
+
+	testWalVerify(t, WalVerifyTestSetup{
+		expectedIntegrityCheck: expectedIntegrityCheck,
+		expectedTimelineCheck:  expectedTimelineCheck,
+		currentWalSegment:      currentSegment,
+		storageSegments:        storageSegments,
+		storageFiles:           map[string]*bytes.Buffer{historyName: historyFile},
+	})
+}
+
+// check that wal-verify correctly reports lost segments
+func TestWalVerify_TwoTimelines_SomeLost(t *testing.T) {
+	storageSegments := []string{
+		"000000050000000000000001",
+		"000000050000000000000002",
+		"000000050000000000000004",
+		"000000050000000000000005",
+		"000000050000000000000006",
+		"000000060000000000000007",
+		"000000060000000000000008",
+	}
+
+	// set switch point to somewhere in the 5th segment
+	switchPointLsn := 5*internal.WalSegmentSize + 100
+	historyContents := fmt.Sprintf("%d\t0/%X\tsome comment...\n\n", 5, switchPointLsn)
+	historyName, historyFile, err := newTimelineHistoryFile(historyContents, 6)
+	// .history file should be stored in wal folder
+	historyName = utility.WalPath + historyName
+	assert.NoError(t, err)
+
+	currentSegmentName := "000000060000000000000009"
+	currentSegment, _ := internal.NewWalSegmentDescription(currentSegmentName)
+
+	expectedIntegrityCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusFailure,
+		Details: internal.IntegrityCheckDetails{
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000001",
+				EndSegment:    "000000050000000000000002",
+				SegmentsCount: 2,
+				Status:        internal.Found,
+			},
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000003",
+				EndSegment:    "000000050000000000000003",
+				SegmentsCount: 1,
+				Status:        internal.Lost,
+			},
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000004",
+				EndSegment:    "000000050000000000000004",
+				SegmentsCount: 1,
+				Status:        internal.Found,
+			},
+			{
+				TimelineId:    6,
+				StartSegment:  "000000060000000000000005",
+				EndSegment:    "000000060000000000000006",
+				SegmentsCount: 2,
+				Status:        internal.ProbablyUploading,
```

Comment:
And missing WAL's on the current timeline is always considered as "ProbablyUploading"?

Reply:
I've added illustration to docs in a1fd3b3